### PR TITLE
Re-enable visibility helper's flanking previews

### PR DESCRIPTION
**Pull request by chrishayesmu**
This change covers:

* Adding configuration to enable flank previews
* Supporting new colors in unit discs and unit flag for flanks; disc colors are configurable also
* Improving performance of flanking indicators to avoid large FPS drops
* Fixing a vis helper bug which could leave some tiles still blocked on loading a save

### CODE REVIEW

**Review by moben — COMMENTED** [review]
Looks pretty good and seems to work, a few comments inline.

I think you also need https://github.com/chrishayesmu/XCOM-LW-CommunityEdition/commit/d2b6fd64b902f72991ad0acbb5591ca6965003c6 from the original sightlines mod or similar, otherwise grenade targeting previews can get stuck in a weird flashing update loop. Easily reproduced when using any grenade and targeting one of your own soldiers.

Could also be a bug in here that makes this case call `ProcessNewPosition` more often than intended.

**Comment by moben** [comment]
We can also discuss https://github.com/chrishayesmu/XCOM-LW-CommunityEdition/commit/d2b6fd64b902f72991ad0acbb5591ca6965003c6 separately if you prefer, it might be better to move that to LWCETacticalVisibility and only do sightline calculations and move if the current action is XGAction_Path :thinking:

File: LWCE_Core/Src/XComLongWarCommunityEdition/Classes/LWCETacticalVisibilityHelper.uc [hunk 1]
```diff
@@ -460,11 +504,12 @@ protected function SetVisibilityMarkers(XGUnit kActiveUnit, XGUnit kHelper)
         }
 
         bFlanked = false;
-        if (bVisible)
+
+        if (bVisible && bShowFlanks && kUnit.GetTeam() == eTeam_Alien)
```
[the remark this answers is not in the document]
Comment (chrishayesmu) [reply LWCE_Core/Src/XComLongWarCommunityEdition/Classes/LWCETacticalVisibilityHelper.uc:508]: That would be nice but I can't really think of a good way to visualize that. Using the same flank indicator as for enemies feels strange. Any ideas?

File: LWCE_Core/Config/DefaultLWCEQualityOfLife.ini [hunk 2]
```diff
@@ -21,8 +27,11 @@ bShowForNeutralUnits=false  ; neutral units; typically this is only civilians on
 ; What color disc to show under units in LoS preview. Choices are
 ; eVDC_None, eVDC_Green, eVDC_Gold, eVDC_Orange, eVDC_Red, and eVDC_White.
 eDiscColorForEnemyUnits=eVDC_None
+eDiscColorForFlankedEnemyUnits=eVDC_None
 eDiscColorForFriendlyUnits=eVDC_None
+eDiscColorForFlankedFriendlyUnits=eVDC_None
 eDiscColorForNeutralUnits=eVDC_None
+eDiscColorForFlankedNeutralUnits=eVDC_None
```
[the remark this answers is not in the document]
Comment (chrishayesmu) [reply LWCE_Core/Config/DefaultLWCEQualityOfLife.ini:34]: Good point, I wasn't thinking and just added this for symmetry.

File: LWCE_Core/Src/XComLongWarCommunityEdition/Classes/LWCETacticalVisibilityHelper.uc [hunk 3]
```diff
@@ -351,6 +391,7 @@ protected function MarkUnit(XGUnit kUnit, bool bVisible, bool bFlanked, bool bIs
 
 protected function MoveHelperUnit(XGUnit kUnit, out Vector vLoc)
 {
+    kUnit.SetLocation(vLoc);
```
[the remark this answers is not in the document]
Comment (chrishayesmu) [reply LWCE_Core/Src/XComLongWarCommunityEdition/Classes/LWCETacticalVisibilityHelper.uc:394]: Honestly I'm not sure. We're using the unit location in our flank check, with `kUnit.IsFlankedByLoc(kHelper.Location)`. I considered just changing that to use the pawn, but then I thought "why shouldn't the unit and pawn be in the same location anyway?" It's really hard to tell whether it matters though.

**Comment by chrishayesmu** [comment]
I think I may experiment with your idea of moving it into `LWCETacticalVisibilityHelper` and see if it works okay from there. Just feels cleaner to me.

**Comment by moben** [comment]
I was testing this a bit and noticed that when loading a tac save, sometimes a tile that an xcom unit occupies is unblocked for movement. I think it's the tile that the first automatically selected soldier is on.

Not sure if it's new bug, it doesn't happen consistently.

**Comment by chrishayesmu** [comment]
> I was testing this a bit and noticed that when loading a tac save, sometimes a tile that an xcom unit occupies is unblocked for movement. I think it's the tile that the first automatically selected soldier is on.
> 
> Not sure if it's new bug, it doesn't happen consistently.

I believe that was from my changes. Should be fixed now, or at least I haven't been able to reproduce it.

Newest commit fixes a bunch of situations where that could occur, and also the grenade throwing issue which you sent me the fix for previously. Notably the `ttc` command can still cause vis helpers to become unsynced from the cursor position, but I'll address that separately, this is large enough for now.

File: LWCE_Core/Src/XComLongWarCommunityEdition/Classes/LWCETacticalVisibilityHelper.uc [hunk 4]
```diff
@@ -254,22 +282,140 @@ protected function InitializeHelpers()
...
+    {
+        if (IsVisHelper(kUnit))
+        {
+            continue;
+        }
+
+        kWorldData.GetTileCoordinatesFromPosition(kUnit.Location, iUnitX, iUnitY, iUnitZ);
+
+        // Our vis helpers don't always quite match unit positions on the Z axis, so add a small tolerance in that dimension
+        if (iUnitX == iCoverHelperX && iUnitY == iCoverHelperY && Abs(iUnitZ - iCoverHelperZ) <= 2)
```
Comment (moben) [inline LWCE_Core/Src/XComLongWarCommunityEdition/Classes/LWCETacticalVisibilityHelper.uc:397]: Should we log a warning in these cases? As far as I understand we never intentionally move the helper to the same tile as another unit.

Also: It's so weird that even the tile coordinates are different, I thought you just meant the unit locations :open_mouth:

**Review by moben — APPROVED** [review]
Looks good and seems to work correctly. Some small suggestions but no blockers imo.

I'm not in love with the whole "try to disentangle two units that we accidentally moved to the same tile" concept, but I don't think we can avoid it as it even happens with normal movement in the base game.

File: LWCE_Core/Src/XComLongWarCommunityEdition/Classes/LWCETacticalVisibilityHelper.uc [hunk 5]
```diff
@@ -56,22 +56,29 @@ const COVER_USING_HELPER_PAWN_TYPE     = 30; // ePawnType_Sectoid
...
 var config EVisDiscColor eDiscColorForEnemyUnits;
+var config EVisDiscColor eDiscColorForFlankedEnemyUnits;
 var config EVisDiscColor eDiscColorForFriendlyUnits;
+var config EVisDiscColor eDiscColorForFlankedFriendlyUnits;
 var config EVisDiscColor eDiscColorForNeutralUnits;
 
-var Vector m_vLastValidCursor;
-var Vector m_vLastValidDestination;
+var XGUnit m_kNonCoverUsingHelper;
+var XGUnit m_kCoverUsingHelper;
```
[the remark this answers is not in the document]
Comment (chrishayesmu) [reply LWCE_Core/Src/XComLongWarCommunityEdition/Classes/LWCETacticalVisibilityHelper.uc:76]: Oh yeah, those were for some console commands. Thanks, will re-protect them.

**Pull request merged**